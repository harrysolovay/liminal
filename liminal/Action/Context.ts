import type { Action } from "./Action.js"
import type { ActorLike } from "../common/ActorLike.js"
import { ActionBase } from "./ActionBase.js"
import type { EventBase } from "./event_common.js"
import type { ActionEvent } from "./ActionEvent.js"
import type { Spec } from "../Spec.js"

export function Context<K extends string, Y extends Action, R = string>(
  key: K,
  system: string,
  implementation: ActorLike<Y, R>,
): Generator<
  Context<{
    Model: never
    Event: never
  }>,
  Awaited<R>
>
export function Context<K extends string, Y extends Action, R = string>(
  key: K,
  implementation: ActorLike<Y, R>,
): Generator<
  Context<{
    Model: never
    Event: never
  }>,
  Awaited<R>
>
export function* Context<Y extends Action, R = string>(
  key: string,
  a0: string | ActorLike,
  a1?: ActorLike,
): Generator<Context, Awaited<R>> {
  return yield ActionBase("Context", {
    key,
    system: typeof a0 === "string" ? a0 : undefined,
    implementation: typeof a0 === "string" ? a1 : a0,
  })
}

export interface Context<S extends Spec = Spec> extends ActionBase<"Context", S> {
  key: string
  system: string | undefined
  implementation?: ActorLike
}

export interface ContextEvent<K extends string = string, E extends ActionEvent = ActionEvent>
  extends EventBase<"Context"> {
  key: K
  inner: E
}
