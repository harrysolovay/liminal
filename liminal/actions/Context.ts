import type { ActionEvent } from "../ActionEvent.ts"
import type { ActionLike } from "../Actor/ActionLike.ts"
import type { ActorLike } from "../Actor/ActorLike.ts"
import { reduce } from "../Actor/reduce.ts"
import { Scope } from "../Scope.ts"
import type { ExtractSpec, Spec } from "../Spec.ts"
import { unwrapDeferred } from "../util/unwrapDeferred.ts"
import { ActionBase, type ActionEventBase, type EnteredEvent, type ExitedEvent } from "./actions_base.ts"

export interface Context<S extends Spec = Spec> extends ActionBase<"context", S> {
  key: keyof any
  implementation: ActorLike
}

export function* context<K extends keyof any, Y extends ActionLike, S extends ExtractSpec<Y>, T = string>(
  key: K,
  implementation: ActorLike<Y, T>,
): Generator<
  Context<{
    Entry: S["Entry"]
    Event: ContextEvent<K, EnteredEvent | S["Event"] | ExitedEvent<T>>
  }>,
  T
> {
  return yield ActionBase("context", {
    key,
    implementation,
    async reduce(scope) {
      const actor = unwrapDeferred(implementation)
      const events = scope.events.child((event) => ({
        type: "context",
        context: key,
        event,
      }))
      events.emit({
        type: "entered",
      })
      const contextScope = await reduce(
        actor,
        new Scope(
          scope.args,
          key,
          events,
          scope.runInfer,
          scope.runEmbed,
        ),
      )
      events.emit({
        type: "exited",
        result: contextScope.result,
      })
      return scope.spread({
        next: contextScope.result,
        children: [...scope.children, {
          type: "context",
          scope: contextScope,
        }],
      })
    },
  })
}

export interface ContextEvent<K extends keyof any = keyof any, E extends ActionEvent = any>
  extends ActionEventBase<"context">
{
  context: K
  event: E
}
