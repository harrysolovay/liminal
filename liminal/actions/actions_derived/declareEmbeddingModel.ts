import { setEmbeddingModel } from "../../actions/setEmbeddingModel.ts"
import type { RunEmbed } from "../../adapters.ts"
import type { JSONKey } from "../../util/JSONKey.ts"
import { declareArg } from "../declareArg.ts"

export function* declareEmbeddingModel<K extends JSONKey>(key: K) {
  const runEmbed = yield* declareArg(key)<RunEmbed>()
  return yield* setEmbeddingModel(key, runEmbed)
}
