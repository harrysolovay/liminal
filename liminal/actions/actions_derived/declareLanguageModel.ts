import { setLanguageModel } from "../../actions/setLanguageModel.ts"
import type { RunInfer } from "../../adapters.ts"
import type { JSONKey } from "../../util/JSONKey.ts"
import { declareArg } from "../declareArg.ts"

export function* declareLanguageModel<K extends JSONKey>(key: K) {
  const runInfer = yield* declareArg(key)<RunInfer>()
  return yield* setLanguageModel(key, runInfer)
}
