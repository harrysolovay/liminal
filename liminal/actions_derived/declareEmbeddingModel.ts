import type { Action } from "../Action.ts"
import { arg } from "../actions/arg.ts"
import { setEmbeddingModel } from "../actions/setEmbeddingModel.ts"
import type { RunEmbed } from "../adapters.ts"
import type { EmbeddingModelSetEvent } from "../events/EmbeddingModelSetEvent.ts"
import type { JSONKey } from "../util/JSONKey.ts"

export function* declareEmbeddingModel<K extends JSONKey>(key: K): Generator<
  Action<"arg" | "set_embedding_model", {
    Entry: [K, RunEmbed]
    Event: EmbeddingModelSetEvent<K>
    Throw: never
  }>,
  void
> {
  const runEmbed = yield* arg(key)<RunEmbed>()
  return yield* setEmbeddingModel(key, runEmbed)
}
