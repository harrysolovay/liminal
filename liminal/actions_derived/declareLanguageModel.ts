import type { Action } from "../Action.ts"
import { arg } from "../actions/arg.ts"
import { setLanguageModel } from "../actions/setLanguageModel.ts"
import type { RunInfer } from "../adapters.ts"
import type { LanguageModelSetEvent } from "../events/LanguageModelSetEvent.ts"
import type { JSONKey } from "../util/JSONKey.ts"

export function* declareLanguageModel<K extends JSONKey>(key: K): Generator<
  Action<"arg" | "set_language_model", {
    Entry: [K, RunInfer]
    Event: LanguageModelSetEvent<K>
    Throw: never
  }>,
  void
> {
  const runInfer = yield* arg(key)<RunInfer>()
  return yield* setLanguageModel(key, runInfer)
}
