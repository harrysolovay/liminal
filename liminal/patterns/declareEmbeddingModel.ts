import type { Action } from "../Action.ts"
import { arg } from "../actions/Arg.ts"
import { setEmbeddingModel } from "../actions/SetEmbeddingModel.ts"
import type { RunEmbed } from "../adapters.ts"
import type { EmbeddingModelSetEvent } from "../events/EmbeddingModelSetEvent.ts"

export function* declareEmbeddingModel<K extends keyof any>(key: K): Generator<
  Action<"arg" | "set_embedding_model", {
    Entry: [K, RunEmbed]
    Event: EmbeddingModelSetEvent<K>
    Throw: never
  }>,
  void
> {
  const runEmbed = yield* arg(key)<RunEmbed>()
  return yield* setEmbeddingModel(key, runEmbed)
}
