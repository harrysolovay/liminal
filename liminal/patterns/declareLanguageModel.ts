import { type Arg, arg } from "../actions/Arg.ts"
import { type RunInfer, type SetLanguageModel, setLanguageModel } from "../actions/SetLanguageModel.ts"
import type { LanguageModelSetEvent } from "../actions/SetLanguageModel.ts"

export function* declareLanguageModel<K extends keyof any>(key: K): Generator<
  | Arg<{
    Field: { [_ in K]: RunInfer }
    Event: never
  }>
  | SetLanguageModel<{
    Field: never
    Event: LanguageModelSetEvent<K>
  }>,
  void
> {
  const runInfer = yield* arg(key)<RunInfer>()
  return yield* setLanguageModel(key, runInfer)
}
