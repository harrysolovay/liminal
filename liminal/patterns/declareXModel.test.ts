import { describe, expect, it } from "bun:test"
import { apply } from "../apply.ts"
import { context } from "../Context/Context.ts"
import { TestEmbeddingModel } from "../testing/TestEmbeddingModel/TestEmbeddingModel.ts"
import { TestLanguageModel } from "../testing/TestLanguageModel/TestLanguageModel.ts"
import { declareEmbeddingModel } from "./declareEmbeddingModel.ts"
import { declareLanguageModel } from "./declareLanguageModel.ts"

describe("Model", () => {
  it("generates the expected event sequence", async () => {
    const scope = await apply(function*() {
      yield* declareLanguageModel("default")
      yield* declareLanguageModel("secondary")
      yield* context("child", function*() {
        yield* declareLanguageModel("child_a")
        yield* declareEmbeddingModel("child_b")
      })
      yield* declareEmbeddingModel("tertiary")
    }, {
      default: TestLanguageModel(),
      secondary: TestLanguageModel(),
      child_a: TestLanguageModel(),
      child_b: TestEmbeddingModel(),
      tertiary: TestEmbeddingModel(),
    }).exec()
    expect(JSON.stringify(scope, null, 2)).toMatchSnapshot()
  })
})
