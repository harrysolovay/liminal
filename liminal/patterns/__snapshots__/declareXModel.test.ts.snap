// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Model generates the expected event sequence 1`] = `
"{
  "events": [
    {
      "type": "entered"
    },
    {
      "type": "language_model_set",
      "key": "default"
    },
    {
      "type": "language_model_set",
      "key": "secondary"
    },
    {
      "type": "embedding_model_set",
      "key": "tertiary"
    },
    {
      "type": "exited"
    }
  ],
  "children": [
    {
      "type": "context",
      "scope": {
        "key": "child",
        "events": [
          {
            "type": "entered"
          },
          {
            "type": "language_model_set",
            "key": "child_a"
          },
          {
            "type": "embedding_model_set",
            "key": "child_b"
          },
          {
            "type": "exited"
          }
        ],
        "children": []
      }
    }
  ]
}"
`;
